File: sql/postgresql/Cpp/PostgreSQLLexerBase.cpp
#include "antlr4-runtime.h"
#include "PostgreSQLLexerBase.h"
#include "PostgreSQLLexer.h"
#include <locale>
#include <codecvt>
#include <cwctype>

PostgreSQLLexerBase::PostgreSQLLexerBase(antlr4::CharStream * input) : antlr4::Lexer(input)
{
    _input = input;
}

void PostgreSQLLexerBase::PushTag()
{
	tags.push(this->getText());
}

bool PostgreSQLLexerBase::IsTag()
{
	return this->getText() == tags.top();
}

void PostgreSQLLexerBase::PopTag()
{
	tags.pop();
}

void PostgreSQLLexerBase::UnterminatedBlockCommentDebugAssert()
{
}

bool PostgreSQLLexerBase::CheckLaMinus()
{
	return this->getInputStream()->LA(1) != '-';
}

bool PostgreSQLLexerBase::CheckLaStar()
{
	return this->getInputStream()->LA(1) != '*';
}

bool PostgreSQLLexerBase::CharIsLetter()
{
	return std::iswalpha(static_cast<char>(this->getInputStream()->LA(-1)));
}

void PostgreSQLLexerBase::HandleNumericFail()
{
	this->getInputStream()->seek(this->getInputStream()->index() - 2);
	this->setType(PostgreSQLLexer::Integral);
}

void PostgreSQLLexerBase::HandleLessLessGreaterGreater()
{
	if (this->getText() == "<<") this->setType(PostgreSQLLexer::LESS_LESS);
	if (this->getText() == ">>") this->setType(PostgreSQLLexer::GREATER_GREATER);
}


char32_t surrogate_to_utf32(char16_t high, char16_t low)
{ 
	return (high << 10) + low - 0x35fdc00; 
}

int toCodePoint(int high, int low)
{
	return surrogate_to_utf32(high, low);
}


bool PostgreSQLLexerBase::CheckIfUtf32Letter()
{
	char high = static_cast<char>(this->getInputStream()->LA(-2));
	char low = static_cast<char>(this->getInputStream()->LA(-1));
	return std::iswalpha(toCodePoint(high, low));
}

bool PostgreSQLLexerBase::IsSemiColon()
{
	return  ';' == (char)this->getInputStream()->LA(1);
}

File: sql/postgresql/Dart/PostgreSQLLexerBase.dart
import 'dart:io';
import 'package:antlr4/antlr4.dart';
import 'PostgreSQLLexer.dart';

abstract class PostgreSQLLexerBase extends Lexer {
  List<String> tags = [];

  PostgreSQLLexerBase(CharStream input) : super(input);

  void PushTag() {
    tags.add(this.text);
  }

  bool IsTag() {
    return tags.isNotEmpty && this.text == tags.last;
  }

  void PopTag() {
    if (tags.isNotEmpty) {
      tags.removeLast();
    }
  }

  void UnterminatedBlockCommentDebugAssert() {
    assert(inputStream.LA(1) == -1); // EOF
  }

  bool CheckLaMinus() {
    return this.inputStream.LA(1) != '-'.codeUnitAt(0);
  }

  bool CheckLaStar() {
    return this.inputStream.LA(1) != '*'.codeUnitAt(0);
  }

bool isLetter(String ch) {
  if (ch.isEmpty) {
    return false;
  }

  int rune = ch.codeUnitAt(0);
  return (rune >= 0x41 && rune <= 0x5A) || (rune >= 0x61 && rune <= 0x7A);
}

  bool CharIsLetter() {
    return isLetter(String.fromCharCode(inputStream.LA(-1)!));
  }

  void HandleNumericFail() {
    inputStream.seek(this.inputStream.index - 2);
    type = PostgreSQLLexer.TOKEN_Integral;
  }

  void HandleLessLessGreaterGreater() {
    if (text == "<<") type = PostgreSQLLexer.TOKEN_LESS_LESS; // Define PostgreSQLLexer constants as needed
    if (text == ">>") type = PostgreSQLLexer.TOKEN_GREATER_GREATER;
  }

  bool CheckIfUtf32Letter() {
    var char32 = String.fromCharCode(inputStream.LA(-2)!) +
        String.fromCharCode(inputStream.LA(-1)!);
    return isLetter(char32);
  }

  bool IsSemiColon() {
    return ';' == String.fromCharCode(inputStream.LA(1)!);
  }

  bool IsLetter(String char) {
    return RegExp(r'^[a-zA-Z]$').hasMatch(char);
  }
}
